### DISCUSSION TITLE
Profile 7 to Profile 8 / Interleaved Bitstream Question

### DISCUSSION THREAD

**Discussion by jebaum** [quietvoid/dovi_tool#287] [Q&A] [answered]
Thanks for the awesome tool, and apologies if this information was already provided in one of the below linked issues and I just didn't comprehend it, or in another issue I missed:

My hardware can play DV profiles 5 and 8, not 7. I have DV profile 7 files with FELs that I want to convert to profile 8. There are several different ways to do this, and I'd like to understand if one is technically "best".

It seems like in a profile 7 file, the RPU.bin is contained within the FEL. I can use `dovi_tool demux -m 5` to separate out the BL and the FEL, extract the RPU from the FEL with `dovi_tool extract-rpu EL.hevc`, and inject the RPU back into the base layer to get a profile 8 file. This new file will be several GBs smaller than the original, since the FEL was effectively removed (in my test cases, the FEL was usually about 4.5gb).

Alternatively, after demuxing like the above, I can mux the `BL.hevc` and `EL.hevc` back together with `dovi_tool mux` instead of extracting `RPU.bin` and injecting it. I'll end up with a profile 8 file that seems to have the FEL still contained in it, based on the fact that the file size is about the same as the original.

So my main question is - should one of these methods be considered "better" than the other? I suppose if I can't visually tell the difference it doesn't really matter but I'd like to understand what's happening, and if it's possible to get the benefits of a FEL in a profile 8 file.
Put concretely, is there a better way to convert a profile 7 file to profile 8 than the following:
```
ffmpeg -i input.mkv -c:v copy -bsf:v hevc_mp4toannexb -f hevc - | dovi_tool -m 5 demux -
dovi_tool mux --bl BL.hevc --el EL.hevc
# use mkvtoolnix or equivalent to replace the hevc video stream in input.mkv with the new 'BL_EL_RPU.hevc' video stream created by the previous command
````

[This issue](https://github.com/quietvoid/dovi_tool/issues/111) is very relevant, but it was created before [release 1.5.0](https://github.com/quietvoid/dovi_tool/issues/128#issuecomment-1114139871) added the feature to the `mux` subcommand for creating BL+EL interleaved bitstreams, so I'm not sure if the answer would be different now.
[This issue](https://github.com/quietvoid/dovi_tool/issues/79) is relevant as well but also predates the improved mux functionality.

**Comment by quietvoid** [comment, marked as answer]
Neither of your methods is correct.

The only correct way to convert to profile 8 is to use mode 2.
Mode 5 is only there as an option for old behaviour and is incorrect on its own.

And if you convert to profile 8, it's preferable to use `extract-rpu` and inject rather than `mux`, since there's no point in keeping the EL bitstream.

As for FEL, there's no way to retain it without transcoding and "merging" the FEL in the video with tools such as [DoViBaker](https://github.com/erazortt/DoViBaker).

**Comment by jebaum** [comment]
Understood, thanks for the info. I'll look at avisynth and dovibaker and play with that, or just live with `extract` + `inject` and `-m 2`.

I'm curious - if there's no point in keeping the EL bitstream without using something like dovibaker, why add the muxing functionality for creating interleaved bitstreams to `dovi_tool`? I'm sure there's a good reason, but I can't think of a use case if the EL bitstream won't get used.

**Reply by quietvoid** [reply]
`mux` allows you to keep the original profile 7 video in a single stream.
It's not supported by many devices but it's more convenient than keeping the whole disc rip.

In the past only MakeMKV and yusesope's Python script could do it, but the script was slow.
Nowadays MKVToolNix also does it by default, but it's still useful to be able to do manually on raw bitstreams.

**Reply by quietvoid** [reply]
Also be aware that not doing the FEL processing can result in incorrect brightness in playback for some titles.
It's required. Or you have to regenerate the metadata to be accurate to the HDR10 BL.

**Reply by jebaum** [reply]
I guess I was probably overthinking this, and in general for folks who don't want to dive deep into the weird fractured world of dolby vision, just using the  `dovi_tool -m 2 convert --discard -` command from the readme is probably the way to go.

Thanks for the info, I didn't know about dovibaker so I have some new things to learn when I have some spare time

**Comment by jebaum** [comment]
In case other people going down this same rabbit hole stumble on this discussion, I'll add a few more useful links:

[macos tool to convert from dv7 to dv8](https://github.com/nekno/DV7toDV8)
the bulk of that tool is just [this shell script](https://github.com/nekno/DV7toDV8/blob/main/src/DV7toDV8.sh), which is simple to read and see what commands they're using (basically the same ones discussed above)

[this forum thread](https://community.firecore.com/t/dolby-vision-profile-7-8-support-ts-mkv-files/19713/1174) is about profile 7 and profile 8 support for a mac video player called "Infuse", but in the thread and in the linked post there's a lot of informative discussion about converting between the two.

[this pastebin](https://pastebin.com/0XABb4Ww) contains a script for scanning your media library and outputting a csv of metadata about the dolby vision profiles of the files it finds

between quietvoid's answers above and all of those links, I think that covers everything anyone needs to know about this topic, unless they're developing things like `dovi_tool`

bonus: there's a long forum discussion [here](https://forum.doom9.org/showthread.php?t=183479&page=45) about the ddvt tool that uses `dovi_tool` under the hood to do profile 7 -> 8 conversion along with a few other things. it's 45 pages and the tool only works on windows so I didn't read all of it, but there's some useful tidbits in the discussion.
